### PR TITLE
add convert2cc command

### PR DESCRIPTION
**Pull request by scgbckbone**
As electrum needs to have ckcc-protocol to communicate with coldcard I thought that it may be a good place to add this functionality. We can then further try to incorporate this into electrum itself. 

Providing this PR as a kind of sanity check, whether this is what you wanted (or at least something similar) and a place to further discuss this functionality. 

I also think that coldcard docs need to be updated after this is implemented so that the `--help` is not the only documentation.

### CODE REVIEW

**Comment by scgbckbone** [comment]
consider this to be working POC that was only tested with ledger/trezor on electrum 4.1.5

File: ckcc/cli.py [hunk 1]
```diff
@@ -1010,4 +1010,105 @@ def get_storage_locker():
 
     click.echo(ls)
 
+
+@main.command('coldcardify')
+@click.argument('file', type=click.File("r+"), required=True)
+@click.option('--outfile', '-o', type=click.File('w'), help="output file path where adjusted wallet file is written")
+@click.option('--no-op', default=False, is_flag=True, help="do not write files instead pretty print to console")
```
Comment (doc-hex) [inline ckcc/cli.py:1017]: rename: "dry-run" and "-n"

File: requirements.txt [hunk 2]
```diff
@@ -10,3 +10,4 @@ click>=6.7
 # required by link-layer encryption in client.py
 ecdsa>=0.13
 pyaes
+pytest
```
Comment (doc-hex) [inline requirements.txt:13]: must not be here; move to new "requirements-test" or "requirements-dev"

Comment (doc-hex) [reply requirements.txt:13]: altho, maybe only devs are using this anyway, thx to setup.py

File: ckcc/cli.py [hunk 3]
```diff
@@ -1010,4 +1010,105 @@ def get_storage_locker():
...
+        fingerprint_missmatch_msg = "Fingerprint missmatch! Is this a correct coldcard/wallet file?"
+        assert xfp == contents[KEYSTORE]["root_fingerprint"], fingerprint_missmatch_msg
+        label = "Coldcard {}".format(xfp)
+        contents[KEYSTORE]["label"] = label
+        # 5. ckcc xpub (master xpub)
+        master_ext_pubkey = dev.send_recv(CCProtocolPacker.get_xpub("m"), timeout=None)
+        contents[KEYSTORE]["ckcc_xpub"] = master_ext_pubkey
+
+    if no_op:
+        pprint.pprint(contents)
```
Comment (doc-hex) [inline ckcc/cli.py:1099]: maybe json.dump(indent=2)?

**Comment by doc-hex** [comment]
- as someone who gets upset once or twice a week about binascii.unhexify, yeah the name is not ideal
- working in-place is usually a bad idea, especially for high-stakes data like this ... IMHO, block in-place, force new file (based on old filename+"coldcard")

**Comment by doc-hex** [comment]
- if this is to become part of Electrum, the main code should be a function, in a new file under ckcc
- organize the cli flags and such as is, but then call the function, which probably would be JSON or objects in and out

**Comment by scgbckbone** [comment]
any name for this in mind  ? I would use some inspiration after 'coldcardify' as you can see XD

**Comment by scgbckbone** [comment]
* removed in-place file handling (you're right smthg can go wrong) - now I only open original file for reading and at most for time needed to read it and I immediately close it afterwrds - new file is created in same location as original with '_cc' suffix
* created new module electrum - where the function to adjust keystore is located (for now in place dict adjustments, but can be changed and probably will be as I haven't looked electrum code thoroughly)

File: ckcc/cli.py [hunk 4]
```diff
@@ -1010,4 +1006,70 @@ def get_storage_locker():
...
+    assert contents["wallet_type"] == "standard", "Not a standard wallet"
+    # below line adjust keystore dict in contents dict (in place)
+    cc_adjust_hww_keystore(contents["keystore"], dev)
+
+    content_str = json.dumps(contents, indent=4)
+    if dry_run:
+        click.echo(content_str)
+    else:
+        try:
+            outfile = outfile or "{}_cc".format(file)
```
Comment (doc-hex) [inline ckcc/cli.py:1064]: need to consider file extension here, which might be `.json`

File: ckcc/cli.py [hunk 5]
```diff
@@ -1010,4 +1006,70 @@ def get_storage_locker():
...
+        with open(file, "r") as f:
+            contents = json.loads(f.read())
+    except Exception as e:
+        click.echo("Failed to load wallet file: {}".format(e))
+        sys.exit(1)
+
+    # for now only standard wallet
+    assert contents["wallet_type"] == "standard", "Not a standard wallet"
+    # below line adjust keystore dict in contents dict (in place)
+    cc_adjust_hww_keystore(contents["keystore"], dev)
```
Comment (doc-hex) [inline ckcc/cli.py:1057]: IIRC multisig wallets have multiple "keystore" entries, but should otherwise be possible to convert

**Comment by doc-hex** [comment]
please add multisig support

**Comment by nvk** [comment]
NACK poetry. 

KISS; Please don’t change/add dependency management. This could break many other projects"

File: ckcc/cli.py [hunk 6]
```diff
@@ -1010,4 +1006,97 @@ def get_storage_locker():
...
+
+keystore_keys = ["derivation", "hw_type", "label", "root_fingerprint", "soft_device_id", "xpub"]
+
+@main.command('coldcardify')
+@click.argument('file', type=click.Path(exists=True), required=True)
+@click.option('--outfile', '-o', type=click.Path(),
+              help="output file path where adjusted wallet file is written. "
+                   "If this is not specified output is written to <original_file>_cc")
+@click.option('--dry-run', '-n', default=False, is_flag=True, help="do not write files instead pretty print to console")
+@click.option('--key', '-k', type=click.Choice(keystore_keys),
```
Comment (scgbckbone) [inline ckcc/cli.py:1018]: key/val is probably too versatile, but I did not want to prohibit this to only hw_type (as one can have multiple from same vendor) or fingerprint...

**Comment by scgbckbone** [comment]
those data were unused and only wasted space tbh, whe have never used the in test - therefore as the only thing we need is keystore/s I removed additional wallet data. My other thinking was that now the test keystores are more readable

**Comment by doc-hex** [comment]
new name: `convert2cc`

**Comment by scgbckbone** [comment]
love it (updated).

Here is the pysecp branch that I mentioned yesterday
https://github.com/scgbckbone/ckcc-protocol/tree/pysecp (build on top of
this https://github.com/scgbckbone/ckcc-protocol/tree/segfault_fixes
because without this branch ckcc is unusable on ubuntu 20)

from the perspective of pysecp256k1 it is built on top of this branch
https://github.com/scgbckbone/python-secp256k1/compare/coldcard_ecdh I
haven't merged yet, as I still have som thinking around it
https://github.com/scgbckbone/python-secp256k1/issues/2

**Comment by scgbckbone** [comment]
* name updated
* test files reverted

**Pull request merged**